### PR TITLE
Restores smooth scrolling for sub-link nav items

### LINKED ISSUE #669: Scrolling to a sub-link item in the nav no longer smooths scrolls
Changes to the focus system broke the smooth scrolling. Before, if you clicked on a "sub-link" item in the nav (The tinier link inside a page) that element would animate to the top.

Now the browser default focus handles this. Therefore either the page doesn't change (if the element is already on screen) or the scroll instantly warps (and the element to focus on is somewhere on the page, not necessarily at the top).

### PR DESCRIPTION
**Pull request by zachberry**
Fixes #669 

* Nav no longer calls scrollIntoView for a sub-link item, instead uses the focus system
* viewer-app scroll to component now does a few hacks to smoothly scroll to an element as before

### CODE REVIEW

**Comment by zachberry** [comment]
Here's what I'm seeing:

**Comment by zachberry** [comment]
> Using the navigation to scroll to questions focuses on the entire question, rather than the "Try question" button. Hitting tab after the question is focused focuses on the "Try question" button, but there is no way for non-sighted users to know this. Perhaps add a screenreader-only dialog to non-revealed questions?
> (Tested with ChromeVox on Chrome and VoiceOver on Safari)

you right. Now it should focus on the content. Also made a change to focus... If you call focusOnContent it'll fall back to simply focusing on the component if that component doesn't have a focusOnContent method.

**Comment by zachberry** [comment]
Ready for another review.

* Removes focus:content, instead, focus:component now always calls the component class' focusOnContent method if it exists, otherwise, just calls focus on the DOM element of the component
* FocusStore/FocusUtil has a new option to specify if the focus event should smooth scroll to the element.
* FocusStore/FocusUtil calls now can take an object of options (previously took a boolean) to specify if the fade effect should happen and if smooth scroll should happen
* clearVisualFocus name changed to clearFadeEffect

**Comment by zachberry** [comment]
Related: https://github.com/ucfopen/Obojobo-Docs/pull/41

**Comment by zachberry** [comment]
Here's the draft I was using to test the UI/UX:

```xml
<?xml version="1.0" encoding="utf-8"?>
<ObojoboDraftDoc>
	<!-- Sample Obojobo Draft Document -->
	<!-- https://ucfopen.github.io/Obojobo-Docs/releases/v3.3.2/authors/ -->
	<Module title="My Obojobo Module Title">

		<Content>
			<Page title="Example content page">
				
				
				
				<ActionButton>
					<textGroup>
						<t>Focus:nav</t>
					</textGroup>
					<triggers>
						<trigger type="onClick">
							<actions>
								<action type="focus:navigation">
								</action>
							</actions>
						</trigger>
					</triggers>
				</ActionButton>
				
				
				
				
				<ActionButton>
					<textGroup>
						<t>Focus:component z</t>
					</textGroup>
					<triggers>
						<trigger type="onClick">
							<actions>
								<action type="focus:component">
									<value id="z" />
								</action>
							</actions>
						</trigger>
					</triggers>
				</ActionButton>
				
				<ActionButton>
					<textGroup>
						<t>Focus:component z (animate)</t>
					</textGroup>
					<triggers>
						<trigger type="onClick">
							<actions>
								<action type="focus:component">
									<value id="z" animateScroll="true" />
								</action>
							</actions>
						</trigger>
					</triggers>
				</ActionButton>
				
				<ActionButton>
					<textGroup>
						<t>Focus:component z (fade)</t>
					</textGroup>
					<triggers>
						<trigger type="onClick">
							<actions>
								<action type="focus:component">
									<value id="z" fade="true" />
								</action>
							</actions>
						</trigger>
					</triggers>
				</ActionButton>
				
				<ActionButton>
					<textGroup>
						<t>Focus:component z (fade, animate)</t>
					</textGroup>
					<triggers>
						<trigger type="onClick">
							<actions>
								<action type="focus:component">
									<value id="z" fade="true" animateScroll="true" />
								</action>
							</actions>
						</trigger>
					</triggers>
				</ActionButton>
				
				
				
				
				
				
				<ActionButton>
					<textGroup>
						<t>Focus:component a</t>
					</textGroup>
					<triggers>
						<trigger type="onClick">
							<actions>
								<action type="focus:component">
									<value id="a" />
								</action>
							</actions>
						</trigger>
					</triggers>
				</ActionButton>
				
				<ActionButton>
					<textGroup>
						<t>Focus:component a (animate)</t>
					</textGroup>
					<triggers>
						<trigger type="onClick">
							<actions>
								<action type="focus:component">
									<value id="a" animateScroll="true" />
								</action>
							</actions>
						</trigger>
					</triggers>
				</ActionButton>
				
				<ActionButton>
					<textGroup>
						<t>Focus:component a (fade)</t>
					</textGroup>
					<triggers>
						<trigger type="onClick">
							<actions>
								<action type="focus:component">
									<value id="a" fade="true" />
								</action>
							</actions>
						</trigger>
					</triggers>
				</ActionButton>
				
				<ActionButton>
					<textGroup>
						<t>Focus:component a (fade, animate)</t>
					</textGroup>
					<triggers>
						<trigger type="onClick">
							<actions>
								<action type="focus:component">
									<value id="a" fade="true" animateScroll="true" />
								</action>
							</actions>
						</trigger>
					</triggers>
				</ActionButton>
				
				
				<ActionButton>
					<textGroup>
						<t>Focus:component q</t>
					</textGroup>
					<triggers>
						<trigger type="onClick">
							<actions>
								<action type="focus:component">
									<value id="q" />
								</action>
							</actions>
						</trigger>
					</triggers>
				</ActionButton>
				
				<ActionButton>
					<textGroup>
						<t>Focus:component q (fade, animate)</t>
					</textGroup>
					<triggers>
						<trigger type="onClick">
							<actions>
								<action type="focus:component">
									<value id="q" fade="true" animateScroll="true" />
								</action>
							</actions>
						</trigger>
					</triggers>
				</ActionButton>
				
				
				
				<h1 id="a">Your Title Here</h1>
				<p>Your content here</p>
				
				<h2 id="b">Your Title Here</h2>
				<p>Your content here</p>
				
				<h2 id="c">Your Title Here</h2>
				<p>Your content here</p>
				
				<h2>Your Title Here</h2>
				<p>Your content here</p>
				
				<Question id="q">
					<h1>Your question here</h1>
					<MCAssessment responseType="pick-one" shuffle="true">
						<MCChoice score="100">
							<MCAnswer>
								<p>Answer 1 text</p>
							</MCAnswer>
							<MCFeedback>
								<p>Optional answer 1 feedback</p>
							</MCFeedback>
						</MCChoice>
						<MCChoice score="0">
							<MCAnswer>
								<p>Answer 2 text</p>
							</MCAnswer>
							<MCFeedback>
								<p>Optional answer 2 feedback</p>
							</MCFeedback>
						</MCChoice>
					</MCAssessment>
					<!-- Optional solution: -->
					<solution>
						<Page>
							<p>Add additional information here</p>
						</Page>
					</solution>
				</Question>
				
				<h2>Your Title Here</h2>
				<p>Your content here</p>
				
				<h2>Your Title Here</h2>
				<p>Your content here</p>
				
				<h2>Your Title Here</h2>
				<p>Your content here</p>
				
				<h2>Your Title Here</h2>
				<p>Your content here</p>
				
				<h2>Your Title Here</h2>
				<p>Your content here</p>
				
				<h2>Your Title Here</h2>
				<p>Your content here</p>
				
				<h2>Your Title Here</h2>
				<p>Your content here</p>
				
				<h2>Your Title Here</h2>
				<p>Your content here</p>
				
				<h2>Your Title Here</h2>
				<p>Your content here</p>
				
				<h2>Your Title Here</h2>
				<p>Your content here</p>
				
				<h2>Your Title Here</h2>
				<p>Your content here</p>
				
				<h2>Your Title Here</h2>
				<p>Your content here</p>
				
				<h2 id="x">Your Title Here</h2>
				<p>Your content here</p>
				
				<h2 id="y">Your Title Here</h2>
				<p>Your content here</p>
				
				<h2 id="z">Your Title Here</h2>
				<p>Your content here</p>
			</Page>
			
			
			
			
			
			
			<Page title="Example content page 2">
				<ActionButton>
					<textGroup>
						<t>Focus:content a</t>
					</textGroup>
					<triggers>
						<trigger type="onClick">
							<actions>
								<action type="focus:content">
									<value id="a2" />
								</action>
							</actions>
						</trigger>
					</triggers>
				</ActionButton>
				
				<ActionButton>
					<textGroup>
						<t>Focus:content q</t>
					</textGroup>
					<triggers>
						<trigger type="onClick">
							<actions>
								<action type="focus:content">
									<value id="q2" />
								</action>
							</actions>
						</trigger>
					</triggers>
				</ActionButton>
				
				<ActionButton>
					<textGroup>
						<t>Focus:content a</t>
					</textGroup>
					<triggers>
						<trigger type="onClick">
							<actions>
								<action type="focus:content">
									<value id="a2" animateScroll="true" />
								</action>
							</actions>
						</trigger>
					</triggers>
				</ActionButton>
				
				<ActionButton>
					<textGroup>
						<t>Focus:content q</t>
					</textGroup>
					<triggers>
						<trigger type="onClick">
							<actions>
								<action type="focus:content">
									<value id="q2" animateScroll="true" />
								</action>
							</actions>
						</trigger>
					</triggers>
				</ActionButton>
				
				
				
				
				
				
				
				
				
				<h1 id="a2">Your Title Here</h1>
				<p>Your content here</p>
				
				<h2 id="b2">Your Title Here</h2>
				<p>Your content here</p>
				
				<h2 id="c2">Your Title Here</h2>
				<p>Your content here</p>
				
				<h2>Your Title Here</h2>
				<p>Your content here</p>
				
				<Question id="q2">
					<h1>Your question here</h1>
					<MCAssessment responseType="pick-one" shuffle="true">
						<MCChoice score="100">
							<MCAnswer>
								<p>Answer 1 text</p>
							</MCAnswer>
							<MCFeedback>
								<p>Optional answer 1 feedback</p>
							</MCFeedback>
						</MCChoice>
						<MCChoice score="0">
							<MCAnswer>
								<p>Answer 2 text</p>
							</MCAnswer>
							<MCFeedback>
								<p>Optional answer 2 feedback</p>
							</MCFeedback>
						</MCChoice>
					</MCAssessment>
					<!-- Optional solution: -->
					<solution>
						<Page>
							<p>Add additional information here</p>
						</Page>
					</solution>
				</Question>
				
				<h2>Your Title Here</h2>
				<p>Your content here</p>
				
				<h2>Your Title Here</h2>
				<p>Your content here</p>
				
				<h2>Your Title Here</h2>
				<p>Your content here</p>
				
				<h2>Your Title Here</h2>
				<p>Your content here</p>
				
				<h2>Your Title Here</h2>
				<p>Your content here</p>
				
				<h2>Your Title Here</h2>
				<p>Your content here</p>
				
				<h2>Your Title Here</h2>
				<p>Your content here</p>
				
				<h2>Your Title Here</h2>
				<p>Your content here</p>
				
				<h2>Your Title Here</h2>
				<p>Your content here</p>
				
				<h2>Your Title Here</h2>
				<p>Your content here</p>
				
				<h2>Your Title Here</h2>
				<p>Your content here</p>
				
				<h2>Your Title Here</h2>
				<p>Your content here</p>
				
				<h2 id="x2">Your Title Here</h2>
				<p>Your content here</p>
				
				<h2 id="y2">Your Title Here</h2>
				<p>Your content here</p>
				
				<h2 id="z2">Your Title Here</h2>
				<p>Your content here</p>
			</Page>
		</Content>

		<Assessment id="my-assessment" attempts="3" title="Assessment" review="never">
			<!-- Assessment intro landing page: -->
			<Page>
				<h1 align="center">Assessment Intro Title</h1>
				<p align="center">Some overview before starting the Assessment.</p>
				<p align="center">You have {{assessment:attemptsRemaining}} attempts remaining.</p>

				<!-- This button will start an assessment attempt: -->
				<ActionButton>
					<textGroup>
						<t>Start Assessment</t>
					</textGroup>
					<triggers>
						<trigger type="onClick">
							<actions>
								<action type="assessment:startAttempt">
									<value id="my-assessment" />
								</action>
							</actions>
						</trigger>
					</triggers>
				</ActionButton>
			</Page>

			<!-- Assessment Question Bank: -->
			<QuestionBank choose="Infinity" select="sequential">
				<!-- Inside this QuestionBank is where you'll put all your questions. -->
				<!-- One is provided as an example: -->

				<Question>
					<!-- Question content: -->
					<h1>Your Question Here</h1>
					<p>More question text here.</p>

					<!-- Multiple choice options: -->
					<MCAssessment shuffle="false" responseType="pick-one">
						<MCChoice score="100">
							<MCAnswer>
								<p>Correct answer</p>
							</MCAnswer>
						</MCChoice>

						<MCChoice score="0">
							<MCAnswer>
								<p>Incorrect answer</p>
							</MCAnswer>
						</MCChoice>
					</MCAssessment>

				</Question>
			</QuestionBank>

			<!-- Post-assessment attempt page: -->
			<scoreActions>
				<!-- You can specify mutliple pages depending on the student's highest Assessment Score. -->
				<!-- Only one is included here: -->
				<scoreAction for="[0,100]">
					<Page>
						<h1 align="center">How did you do?</h1>
						<p align="center">You have {{assessment:attemptsRemaining}} attempts remaining.</p>

						<!-- This button will start another assessment attempt: -->
						<ActionButton align="center">
							<textGroup>
								<t>Retry Assessment</t>
							</textGroup>
							<triggers>
								<trigger type="onClick">
									<actions>
										<action type="assessment:startAttempt">
											<value id="my-assessment" />
										</action>
									</actions>
								</trigger>
							</triggers>
						</ActionButton>
					</Page>
				</scoreAction>
			</scoreActions>
		</Assessment>

	</Module>
</ObojoboDraftDoc>

```

File: packages/obojobo-document-engine/src/scripts/viewer/stores/focus-store.js [hunk 1]
```diff
@@ -8,22 +8,24 @@ It doesn't modify any DOM focus and relies on the ViewerApp or OboNode component
 Additionally you can visually focus a component which will fade away all content except the
 component to focus on.
 
-The state has three values: `type`, `target` and `visualFocusTarget`
+The state has three values: `type`, `target`, `animateScroll` and `visualFocusTarget`
```
Comment (iturgeon) [inline packages/obojobo-document-engine/src/scripts/viewer/stores/focus-store.js]:
```suggestion
The state has four possible values: `type`, `target`, `animateScroll` and `visualFocusTarget`
```

**Comment by iturgeon** [comment]
We should try to get clicking on the page title to scroll to the top of the page.  If it's too much work, spin it off into an issue

**Comment by iturgeon** [comment]
I'm going to sit on this review till after split makes it in - the conflict resolution will require additional commits.

**Pull request merged**